src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -74,77 +71,108 @@ ov::Tensor Bank::get(int64_t uid, const std::string& device) {
-            allocated_tensor = ov::make_tensor(remote_tensor);
-            device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = allocated_tensor;
-            dev_guard.unlock();  // Unlock the guard, map update is done - copy can continue in parallel
+void Bank::evaluate_cpu(Bank::DeviceBank& device_bank, const std::vector<LazyTensor>& to_process) {
+    // Note: not locking here. This is a private function, so Bank should handle the locks around it
+    // as we lock in evaluate_and_allocate() now.
+    ov::parallel_for(to_process.size(), [&](std::size_t idx) {
+        const auto& lt = to_process[idx];
+        auto iter_device_registered = device_bank.registered_tensors.find(lt);
+        NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                    "Tensor should be registered first!");
+        auto uid = iter_device_registered->second;
+        device_bank.storage.at(uid).tensor = lt.eval();
+        const_cast<LazyTensor&>(lt).detach();
+    });
+}
 
-            transformed_tensor.copy_to(allocated_tensor);
+// Note: there are no locks in this function's parallel_for
+// At this point all the LazyTensor->Tensor pairs in the map are already
+// allocated and there are no conflicting reads/writes since all the tensors are unique.
+
+// FIXME: this whole flow could be improved for the same bank
+// processing from different threads. We could separate LazyTensors
+// evaluation and the bank access. But it requires additional rework.
+void Bank::evaluate_and_allocate_on_device(Bank::DeviceBank& device_bank,
+                                           const std::vector<LazyTensor>& to_process,
+                                           const std::string& device) {
+    // Note: not locking here. This is a private function, so Bank should handle the locks around it
+    // as we lock in evaluate_and_allocate() now.
+    std::vector<TensorToAllocate> uids_to_allocated;
+    uids_to_allocated.reserve(uid_count);
+
+    for (const auto& lt : to_process) {
+        auto iter_device_registered = device_bank.registered_tensors.find(lt);
+        NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                    "Tensor should be registered first!");
+        auto uid = iter_device_registered->second;
+        uids_to_allocated.push_back({lt.eval_meta(), ov::Tensor(), uid});
+    }
+    // Sort by UIDs, lowest first
+    std::sort(uids_to_allocated.begin(), uids_to_allocated.end());
 
-            // Detach the evaluated LazyTensor from its memory here - when it is 100%
-            // not needed anymore (transformations, if any, and copies are done)
-            // Note: this is the non-CPU path!
-            const_cast<LazyTensor&>(lt).detach();
-        });
+    // Allocate memory sequentially - in order of UID
+    auto remote_ctx = m_core->get_default_context(device)._ptr;
```

Comment:
wait a sec, why do I see `get_default_context` here again? Did you confirm it will be the same context as before? What was the reason to cache them in the earlier versions of this PR?